homeassistant/components/enocean/config_flow.py

```diff
@@ -0,0 +1,109 @@
+    """Handle the enOcean config flows."""
+
+    VERSION = 1
+    MANUAL_PATH_VALUE = "Custom path"
+    CONNECTION_CLASS = CONNECTION_ENOCEAN
+
+    def __init__(self):
+        """Initialize the EnOcean config flow."""
+        self.dongle_path = None
+        self.discovery_info = None
+
+    async def async_step_init(self, user_input=None):
+        """Needed in order to not require re-translation of strings."""
+        return await self.async_step_user(user_input)
+
+    async def async_step_user(self, user_input=None):
+        """Handle an EnOcean config flow start.
+
+        This flow will be used when the user sets up the ENOcean integration.
+        It uses a state machine to walk through the different steps of the
+        configuration.
+        """
+        if self._async_current_entries() or DOMAIN in self.hass.data:
+            return self.async_abort(reason="dongle_already_configured")
+
+        if user_input is None:
+            config_state = DongleSetupStates.SELECT_AUTO
+        elif user_input[CONF_DEVICE] == self.MANUAL_PATH_VALUE:
+            config_state = DongleSetupStates.SELECT_MANUAL
+        else:
+            config_state = DongleSetupStates.VALIDATE
+
+        return await self.set_dongle_setup_state(config_state, user_input)
+
+    async def set_dongle_setup_state(self, new_state, user_input=None, errors=None):
+        """Change the current state of the dongle setup state machine (re-entrant method)."""
+        if new_state == DongleSetupStates.SELECT_MANUAL:
+            LOGGER.debug("Config step: select manual with errors=%s", errors)
+            default_value = None
+            if (
+                user_input is not None
+                and CONF_DEVICE in user_input
+                and user_input[CONF_DEVICE] != self.MANUAL_PATH_VALUE
+            ):
+                default_value = user_input[CONF_DEVICE]
+            return self.async_show_form(
+                step_id="init",
+                data_schema=vol.Schema(
+                    {vol.Required(CONF_DEVICE, default=default_value): str}
+                ),
+                errors=errors,
+            )
+
+        if new_state == DongleSetupStates.SELECT_AUTO:
+            LOGGER.debug("Config step: select auto")
+            bridges = EnOceanDongle.detect()
+            if len(bridges) == 0:
+                return await self.set_dongle_setup_state(
+                    DongleSetupStates.SELECT_MANUAL, user_input
+                )
+
+            bridges.append(self.MANUAL_PATH_VALUE)
+            return self.async_show_form(
+                step_id="init",
+                data_schema=vol.Schema({vol.Required(CONF_DEVICE): vol.In(bridges)}),
+            )
+
+        if new_state == DongleSetupStates.VALIDATE:
+            LOGGER.debug("Config step: validate")
+            dongle_path = user_input[CONF_DEVICE]
+            if path.exists(dongle_path) and not path.isdir(dongle_path):
```

Comment:
I/O. Please use the executor. Probably break out a separate validation function that we can schedule on the executor.

Before creating the entry we also want to check that we can connect to the device.

Reply:
Done.
I added a method in EnOceanDongle to validate the path, that I call using async_add_executor_job.